packages/expo-router/src/loadStaticParamsAsync.ts

```diff
@@ -157,25 +87,83 @@ async function loadStaticParamsRecursive(
         dynamic: null,
         route: parsedRoute,
-        children: uniqBy(
-          (
-            await recurseAndFlattenNodes(
-              [...route.children],
-              {
-                ...props,
-                parentParams: {
-                  ...props.parentParams,
-                  ...staticParams,
-                },
-              },
-              loadStaticParamsRecursive
-            )
-          ).flat(),
-          (i) => i.route
-        ),
+        children: dynamicChildren,
       };
     })
   );
 
   return [route, ...generatedRoutes];
 }
+
+/** lodash.uniqBy */
+function uniqBy<T>(array: T[], key: (item: T) => string): T[] {
+  const seen: { [key: string]: boolean } = {};
+  return array.filter((item) => {
+    const k = key(item);
+    if (seen[k]) {
+      return false;
+    }
+    seen[k] = true;
+    return true;
+  });
+}
+
+function assertStaticParamsType(params: any): asserts params is Record<string, string | string[]> {
+  if (!Array.isArray(params)) {
+    throw new Error(`generateStaticParams() must return an array of params, received ${params}`);
+  }
+}
+
+function assertStaticParams(route: RouteNode, params: Record<string, string | string[]>) {
+  const matches = route.dynamic!.every((dynamic) => {
+    const value = params[dynamic.name];
+    return value !== undefined && value !== null;
+  });
+  if (!matches) {
+    throw new Error(
+      `generateStaticParams() must return an array of params that match the dynamic route. Received ${JSON.stringify(
+        params
+      )}`
+    );
+  }
+
+  const validateSingleParam = (
+    dynamic: DynamicConvention,
+    value: any,
+    allowMultipleSegments?: boolean
+  ) => {
+    if (typeof value !== 'string') {
+      throw new Error(
+        `generateStaticParams() for route "${route.contextKey}" expected param "${
+          dynamic.name
+        }" to be of type string, instead found "${typeof value}" while parsing "${value}".`
+      );
+    }
+    const parts = value.split('/').filter(Boolean);
+    if (parts.length > 1 && !allowMultipleSegments) {
+      throw new Error(
+        `generateStaticParams() for route "${route.contextKey}" expected param "${dynamic.name}" to not contain "/" (multiple segments) while parsing "${value}".`
+      );
+    }
+    if (parts.length === 0) {
+      throw new Error(
+        `generateStaticParams() for route "${route.contextKey}" expected param "${dynamic.name}" not to be empty while parsing "${value}".`
+      );
+    }
+  };
+
+  route.dynamic!.forEach((dynamic) => {
+    const value = params[dynamic.name];
+    if (dynamic.deep) {
+      // TODO: We could split strings by `/` and use that too.
+      if (!Array.isArray(value)) {
+        validateSingleParam(dynamic, value, true);
+      } else {
+        validateSingleParam(dynamic, value.filter(Boolean).join('/'), true);
+      }
+    } else {
+      validateSingleParam(dynamic, value);
+    }
+    return value !== undefined && value !== null;
```

Comment:
I'm confused by this last line, its not doing anything

Reply:
good catch